lib: smf: Add State Machine Framework (#38088)
zephyrproject-rtos/zephyr: doc/guides/smf/index.rst

.. _smf:

State Machine Framework
#######################

Overview
========

The State Machine Framework (SMF) is an application agnostic framework that
provides an easy way for developers to integrate state machines into their

> application. The framework can be added to any project by including the
> CONFIG_SMF=y into the projects prj.conf file.

Reviewer: ```suggestion
application. The framework can be added to any project by enabling the
:kconfig:`CONFIG_SMF` option.
```

Author: Done

State Creation
==============

A state is represented by three functions, where one function implements the
Entry actions, another function implements the Run actions, and the last
function implements the Exit actions. The prototype for these functions is as

> follows: **void funct(void *obj)**, where the **obj** parameter is a user

Reviewer: Please read through the documentation guidelines and make adjustments: https://docs.zephyrproject.org/latest/contribute/documentation/index.html

E.g. this should be using double backticks for inline code literals as follows:

```
follows: ``void funct(void *obj)`` [...]
```

Author: Done

defined structure that has the state machine context, **struct smf_ctx**, as
its first member. For example::

	struct user_object {
 		struct smf_ctx ctx;
 		/* All User Defined Data Follows */
	};

The **struct smf_ctx** member must be first because the state machine
framework's functions casts the user defined object to the **struct smf_ctx**
type with the following macro: **SMF_CTX(o)**

For example instead of doing this **(struct smf_ctx *)&user_obj**, you could
use **SMF_CTX(user_obj)**.

The following macros can be used for easy state creation:
    - HSM_CREATE_MSG_STATE
    - HSM_CREATE_STATE
    - FSM_CREATE_MSG_STATE
    - FSM_CREATE_BASIC_STATE

**HSM_CREATE_MSG_STATE(me, mx, e, r, x, p)** is used to create a hierarchical
state that can have an optional parent state and prints a message in the
optional Entry action and a message in the optional Exit action. The macro has

> the following parameters:

Reviewer 2: Note that the parameters are now out of date.  Probably better to document the parameters in the code only and just provide a summary of the macro and it's intended use here.

Reviewer: I would suggest looking into the `:c:macro:` role and using it as needed to link to the API docs for this instead.

Author: Thanks for the suggestion

added to the projects **prj.conf** file. The use of **FSM_CREATE_MSG_STATE** or
**FSM_CREATE_STATE** when the **CONFIG_SMF_FLAT** config option isn't defined,
results in a build error. If the **CONFIG_SMF_FLAT=y** config option is defined
and **HSM_CREATE_STATE** or **HSM_CREATE_MSG_STATE** is used, results in a build
error.

State Machine Creation
======================

A state machine is created by defining a table of states that's indexed by an
enum. For example::

	enum demo_state { S0, S1, S2 };

	const struct smf_state demo_states {
		[S0] = FSM_CREATE_STATE(s0_entry, s0_run, s0_exit),
		[S1] = FSM_CREATE_STATE(s1_entry, s1_run, s1_exit),
		[S2] = FSM_CREATE_STATE(s2_entry, s2_run, s2_exit)
	};

To set the initial state or to transition from one state to another, the
**set_state** function is used and it has the following prototype:
**void set_state(smf_ctx *ctx, smf_state *state)**

**NOTE:** While the state machine is running, set_state should only be called
from the Entry and Run functions. Calling set_state from the Exit functions
doesn't make sense and will generate a warning.

State Machine Execution

> =======================
> To run the state machine, the **run_state** function should be called in some

Reviewer 3: ```suggestion
=======================

To run the state machine, the **run_state** function should be called in some
```

Author: Done

application dependent way. The function has the following prototype:
**void run_state(smf_ctx *ctx)**

Flat State Machine Example

> ==========================
> This example turns the following state diagram into code using the SMF, where

Reviewer 3: ```suggestion
==========================

This example turns the following state diagram into code using the SMF, where
```

Author: Done

the initial state is S0.

> .. figure:: img/flat.png

Reviewer 3: note: you may want to use `graphviz` diagrams, they are natively supported in the documentation (render to SVG and are better viewed in both light/dark modes)

Author: Yes, the image is much better. Thanks.

	static const struct smf_state demo_states[] {
		[S0] = FSM_CREATE_STATE(s0_entry, s0_run, s0_exit),
		[S1] = FSM_CREATE_STATE(s1_entry, s1_run, s1_exit),
		[S2] = FSM_CREATE_STATE(s2_entry, s2_run, s2_exit),
	};

	void main(void)
	{
		/* Set initial state */
		set_stat(SMF_CTX(s_obj), &demo_states[S0]);

		/* Run the state machine */
		while(1) {
			run_state(SMF_CTX(s_obj));
			k_msleep(1000);
		}
	}

Hierarchical State Machine Example
==================================
This example turns the following state diagram into code using the SMF, where
S0 and S1 share a parent state and S0 is the initial state.

.. figure:: img/hierarchical.png
    :align: center
    :alt: Hierarchial state machine diagram

Code::

	#include <smf.h>

	/* Forward declaration of state table */
	static const struct smf_state demo_states[];

	/* List of demo states */
	enum demo_state { PARENT, S0, S1, S2 };

	/* User defined object */
	struct s_object {
		/* This must be first */
		struct smf_ctx ctx;

		/* Other state specific data add here */
	} s_obj;

	/* Parent State */
	static void parent_entry(void *o) {}
	static void parent_run(void *o) {}
	static void parent_exit(void *o) {}

	/* State S0 */

> 	static void s0_entry(void *o) {}

Reviewer 3: would it be possible to allow `NULL` entry/exit callbacks if they are a no-op?

Author: I made some NULL and added comments to make things clear.

	static void s0_run(void *o)
	{
		set_state(SMF_CTX(s_obj), &demo_states[S1]);
	}
	static void s0_exit(void *o) {}

	/* State S1 */
	static void s1_entry(void *o) {}
	static void s1_run(void *o)
	{
		set_state(SMF_CTX(s_obj), &demo_states[S2]);
	}
	static void s1_exit(void *o) {}

	/* State S2 */
	static void s2_entry(void *o) {}
	static void s2_run(void *o)
	{
		set_state(SMF_CTX(s_obj), &demo_states[S0]);
	}
	static void s2_exit(void *o) {}

	/* Populate state table */
	static const struct smf_state demo_states[] {
		[PARENT] = HSM_CREATE_STATE(parent_entry, parent_run, parent_exit, NULL);
		[S0] = HSM_CREATE_STATE(s0_entry, s0_run, s0_exit, &demo_states[PARENT]),
		[S1] = HSM_CREATE_STATE(s1_entry, s1_run, s1_exit, &demo_states[PARENT]),
		[S2] = HSM_CREATE_STATE(s2_entry, s2_run, s2_exit),
	};

	void main(void)
	{
		/* Set initial state */
		set_stat(SMF_CTX(s_obj), &demo_states[S0]);

		/* Run the state machine */
		while(1) {

> 			run_state(SMF_CTX(s_obj));

Reviewer 3: should error handling be allowed?

Author: I added a way to return an error and terminate the state machine

			k_msleep(1000);
		}
	}

> When designing hierarchical state machines, the following should be considered\:

Reviewer 3: ```suggestion
When designing hierarchical state machines, the following should be considered:

```

Author: Done

>     - Ancestor entry actions are executed before the sibling entry actions. For

Reviewer 3: ```suggestion
- Ancestor entry actions are executed before the sibling entry actions. For
```

Author: Done

smf_run_state if it returns a non-zero value. The function has the following
prototype: ``int32_t smf_run_state(smf_ctx *ctx)``

State Machine Termination
=========================

To terminate the state machine, the ``smf_terminate`` function should be
called. It can be called from the entry, run, or exit action. The function
takes a non-zero user defined value that's returned by the ``smf_run_state``
function. The function has the following prototype:
``int32_t smf_terminate(smf_ctx *ctx, int32_t val)``

Flat State Machine Example
==========================

This example turns the following state diagram into code using the SMF, where
the initial state is S0.

.. figure:: img/flat.svg
    :align: center
    :alt: Flat state machine diagram

Code::

	#include <smf.h>

	/* Forward declaration of state table */
	static const struct smf_state demo_states[];

	/* List of demo states */
	enum demo_state { S0, S1, S2 };

	/* User defined object */
	struct s_object {
		/* This must be first */
		struct smf_ctx ctx;

		/* Other state specific data add here */
	} s_obj;

	/* State S0 */
	static void s0_entry(void *o)
	{
		/* Do something */
	}
	static void s0_run(void *o)
	{

> 		smf_set_state(SMF_CTX(s_obj), &demo_states[S1]);

Reviewer 4: @sambhurst I think & is needed, same as in other SMF_CTX calls

Author: Done

	}
	static void s0_exit(void *o)
	{
		/* Do something */
	}

	/* State S1 */
	static void s1_run(void *o)
	{
		smf_set_state(SMF_CTX(s_obj), &demo_states[S2]);
	}
	static void s1_exit(void *o)
	{
		/* Do something */
	}

	/* State S2 */
	static void s2_entry(void *o)
	{
		/* Do something */
	}
	static void s2_run(void *o)
	{
		smf_set_state(SMF_CTX(s_obj), &demo_states[S0]);
	}

	/* Populate state table */

> 	static const struct smf_state demo_states[] {

Reviewer 4: "`=`" is missing:
`static const struct smf_state demo_states[] = `

Author: Done

				break;
			}
			k_msleep(1000);
		}
	}

Hierarchical State Machine Example
==================================

This example turns the following state diagram into code using the SMF, where
S0 and S1 share a parent state and S0 is the initial state.

.. figure:: img/hierarchical.svg
    :align: center
    :alt: Hierarchial state machine diagram

Code::

	#include <smf.h>

	/* Forward declaration of state table */
	static const struct smf_state demo_states[];

	/* List of demo states */
	enum demo_state { PARENT, S0, S1, S2 };

	/* User defined object */
	struct s_object {
		/* This must be first */
		struct smf_ctx ctx;

		/* Other state specific data add here */
	} s_obj;

	/* Parent State */
	static void parent_entry(void *o)
	{
		/* Do something */
	}
	static void parent_exit(void *o)
	{
		/* Do something */
	}

	/* State S0 */
	static void s0_run(void *o)
	{
		smf_set_state(SMF_CTX(s_obj), &demo_states[S1]);
	}

	/* State S1 */
	static void s1_run(void *o)
	{
		smf_set_state(SMF_CTX(s_obj), &demo_states[S2]);
	}

	/* State S2 */
	static void s2_run(void *o)
	{
		smf_set_state(SMF_CTX(s_obj), &demo_states[S0]);
	}

	/* Populate state table */

> 	static const struct smf_state demo_states[] {

Reviewer 4: I guess also = missing

Author: Done

Author: Done

		/* Parent state does not have a run action */

> 		[PARENT] = HSM_CREATE_STATE(parent_entry, NULL, parent_exit, NULL);

Reviewer 4: Should be , not ;

Author: Done

		/* Child states do not have entry or exit actions */
		[S0] = HSM_CREATE_STATE(NULL, s0_run, NULL, &demo_states[PARENT]),
		[S1] = HSM_CREATE_STATE(NULL, s1_run, NULL, &demo_states[PARENT]),
		/* State S2 do ot have entry or exit actions and no parent */
		[S2] = HSM_CREATE_STATE(NULL, s2_run, NULL, NULL),
	};

	void main(void)
	{
		int32_t ret;

		/* Set initial state */
		smf_set_state(SMF_CTX(s_obj), &demo_states[S0]);

		/* Run the state machine */
		while(1) {
			/* State machine terminates if a non-zero value is returned */

> 			ret = smf_run_state(SMF_CTX(s_obj);

Reviewer 4: missing )

Author: Done
